clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/SafeRetry.groovy

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2016 Google, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.clouddriver.google.deploy
+
+import com.google.api.client.googleapis.json.GoogleJsonResponseException
+import com.google.common.annotations.VisibleForTesting
+import com.netflix.spinnaker.clouddriver.data.task.Task
+import com.netflix.spinnaker.clouddriver.google.deploy.exception.GoogleOperationException
+import groovy.util.logging.Slf4j
+
+import java.util.concurrent.TimeUnit
+
+@Slf4j
+class SafeRetry<T> {
+
+  @VisibleForTesting
+  static long SAFE_RETRY_INTERVAL_MILLIS = TimeUnit.SECONDS.toMillis(10)
+
+  /**
+   * Retry a GCP operation if it fails. Treat any error codes in successfulErrorCodes as success.
+   *
+   * @param operation - The GCP operation.
+   * @param action - String describing the GCP operation.
+   * @param resource - Resource we are operating on.
+   * @param task - Spinnaker task. Can be null.
+   * @param phase
+   * @param retryCodes - GoogleJsonResponseException codes we retry on.
+   * @param successfulErrorCodes - GoogleJsonException codes we treat as success.
+   *
+   * @return Object of type T returned from the operation.
+   */
+  public static T doRetry(Closure<T> operation,
+                          String action,
+                          String resource,
+                          Task task,
+                          String phase,
+                          List<Integer> retryCodes,
+                          List<Integer> successfulErrorCodes) {
+    try {
+      task?.updateStatus phase, "Attempting $action of $resource..."
+      return operation()
+    } catch (GoogleJsonResponseException | SocketTimeoutException _) {
+      log.warn "Initial $action of $resource failed, retrying..."
+
+      int tries = 1
+      Exception lastSeenException = null
+      while (tries < 10) { // Retry 10 times.
+        try {
+          tries++
+          sleep(SAFE_RETRY_INTERVAL_MILLIS) // Sleep for some interval between attempts.
+          log.warn "$action $resource attempt #$tries..."
+          return operation()
+          log.warn "$action $resource attempt #$tries succeeded"
+          return
```

Comment:
Is this `return` statement (and the logging call above it) reachable?

Looks like you probably want to capture the result instead of just using `return operation()` and then return the captured result here.